File: Math/Russain Peasants Algorithm/C++/RussianPeasantsAlgorithm.cpp
#include <iostream>
using namespace std;

int multiply(int a, int b)
{
	int prod = 0;
	while (b > 0)
	{
		if (b % 2 != 0)
		{
 			prod += a;
		}
		a = a << 1; // multiply by 2
		b = b >> 1; // divide by 2
	}

	return prod;
}

int main()
{
	int a, b;
	cin >> a >> b;
	cout << multiply(a, b) << endl;

	return 0;
}

File: Math/Russain Peasants Algorithm/python/russian-peasants-algorithm.py
# Algorithm to find the product of two numbers in logarithmic time
def multiply(a,b):
    if a > b:
        left = a
        right = b
    else:
        left = b
        right = a
    
    prod = 0
    
    while left > 0: # Loop till you reach 1
        if left%2 != 0:  # Add the value to product variable if the left column value is odd
            prod = prod + right
        
        left = left >> 1 # Right shift
        right = right << 1 # Left Shift

    return prod
    
print(multiply(127,68))
